### PR TITLE
core/compiler_hints: add likely() / unlikely() hints

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

The compiler can optimize the code better if it knows whether a condition is likely taken / not taken.

This is usually something the developer should not bother with, but for things like `assert()` where the condition should *always* be true, we can tell the compiler to optimize that path.

### Testing procedure

No logic changes, CI will ensure that all our compilers know about the builtin. 

### Issues/PRs references

https://github.com/RIOT-OS/RIOT/pull/19155#issuecomment-1384094781

### CODE REVIEW

File: core/lib/include/compiler_hints.h [hunk 1]
```diff
@@ -130,6 +132,24 @@ extern "C" {
...
 #endif
 
+/**
+ * @brief   Hint to the compiler that the condition @p x is likely taken
+ *
+ * @param[in] x     Condition that is likely taken
+ *
+ * @return result of @p x
+ */
+#define likely(x)       __builtin_expect((uintptr_t)(x),1)
```
Comment (maribu) [inline core/lib/include/compiler_hints.h:142]:
```suggestion
#define likely(x)       __builtin_expect((uintptr_t)(x), 1)
```

File: core/lib/include/compiler_hints.h [hunk 2]
```diff
@@ -130,6 +132,24 @@ extern "C" {
...
+
+/**
+ * @brief   Hint to the compiler that the condition @p x is likely not taken
+ *
+ * @param[in] x     Condition that is unlikely
+ *
+ * @return result of @p x
+ */
+#define unlikely(x)     __builtin_expect((uintptr_t)(x),0)
```
Comment (maribu) [inline core/lib/include/compiler_hints.h:151]:
```suggestion
#define unlikely(x)     __builtin_expect((uintptr_t)(x), 0)
```

**Review by maribu — APPROVED** [review]
lgtm and for `assert()` it definitely is worth the additional maintenance overhead.

See two inline nitpicks

**Comment by kfessel** [comment]
likely and unlikely will not do the same as changin the optimization level since doing that kind of reordering might enlarge the build result (but still be faster since the hotpath is shorter) 

[godbolt](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:61,endLineNumber:3,positionColumn:61,positionLineNumber:3,selectionStartColumn:61,selectionStartLineNumber:3,startColumn:61,startLineNumber:3),source:'%23include+%3Cstdint.h%3E%0A%0A%23define+assert(cond)+(__builtin_expect((uintptr_t)((cond)),1)+%3F+(void)0+:+_assert_panic())%0A%0Atypedef+struct+%7B%0A++++uint8_t+blockers%5B4%5D%3B+++++/**%3C+number+of+blockers+for+the+mode+*/%0A%7D+__attribute__+((aligned+(8)))+pm_blocker_t%3B%0A%0Astatic+pm_blocker_t+pm_blocker+%3D+%7B+.blockers+%3D+%7B1,1,1,0%7D+%7D%3B%0A//__attribute__((optimize(3)))%0A%0Avoid+pm_block(unsigned+mode)%0A%7B%0A++++//unsigned+state+%3D+irq_disable()%3B%0A++++assert(pm_blocker.blockers%5Bmode%5D+!!%3D+255)%3B%0A++++pm_blocker.blockers%5Bmode%5D%2B%2B%3B%0A++++//irq_restore(state)%3B%0A%7D'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:50,l:'4',m:100,n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:carm1130,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:2,lang:___c,libs:!(),options:'-mcpu%3Dcortex-m0+-Os+-mfloat-abi%3Dsoft+',selection:(endColumn:26,endLineNumber:15,positionColumn:26,positionLineNumber:15,selectionStartColumn:26,selectionStartLineNumber:15,startColumn:26,startLineNumber:15),source:1),l:'5',n:'0',o:'+ARM+gcc+11.3+(linux)+(Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

**Comment by kfessel** [comment]
I think likely does not provide extra information to the compiler since the `__attribute__((noreturn))` already makes clang behave as if that case is unlikely. 
clang does the reordering if the condition is marked likely or the panic() is marked NORETURN
gcc does make the reodering with one of the makers (noreturn or likely) but not if Os is used only if O3/1/2

**Comment by benpicco** [comment]
bors merge

File: core/lib/include/assert.h [hunk 3]
```diff
@@ -22,6 +22,8 @@
 #ifndef ASSERT_H
 #define ASSERT_H
 
+#include "compiler_hints.h"
```
Comment (kfessel) [inline core/lib/include/assert.h:25]:
```suggestion
#define __likely(x)       __builtin_expect((uintptr_t)(x), 1)
```

**Review by kfessel — CHANGES_REQUESTED** [review]
since assert.h implements a standard header it should not provide defines that are not part of that standard header

**Comment by kfessel** [comment]
bors cancel

File: core/lib/include/assert.h (hunk 3 above)
Comment (benpicco) [reply core/lib/include/assert.h]: I added the custom macro instead of `compiler_hints.h`

**Review by kfessel — APPROVED** [review]

**Comment by benpicco** [comment]
bors merge

**Review by maribu — APPROVED** [review]
bors merge

**Comment by maribu** [comment]
bors retry

**Comment by maribu** [comment]
bors cancel
bors merge

**Review by maribu — APPROVED** [review]

**Comment by maribu** [comment]
bors merge

**Comment by maribu** [comment]
bors merge --force

**Comment by maribu** [comment]
bors merge

**Comment by maribu** [comment]
bors retry

**Comment by benpicco** [comment]
bors merge

**Comment by benpicco** [comment]
bors merge

**Comment by kfessel** [comment]
bors status

**Comment by kfessel** [comment]
bors cancel

**Comment by maribu** [comment]
> bors Pending — Waiting in queue 

:tada: So in 3h - 5h it will get merged :)

**Pull request merged**